File: app/src/main/java/com/botijasoftware/utils/Texture.java
package com.botijasoftware.utils;

public class Texture {
	
	public Texture(int id, int w, int h) {
		mID = id;
		width = w;
		height = h;
		textCoords = new TextureCoords();
	}
	
	public Texture(int id, int w, int h, TextureCoords tc) {
		mID = id;
		width = w;
		height = h;
		textCoords = tc;		
	}
	
	public int getID() {
		return mID;
	}
	
	public int getWidth() {
		return width;
	}
	
	public int getHeight() {
		return height;
	}
	
	public TextureCoords getTextureCoords() {
		return textCoords;
	}

    /*public Texture clone() {

        return new Texture(mID, width, height, textCoords.clone());

    }*/

    public TextureRegion subTexture(float S0, float T0, float S1, float T1) {
        return new TextureRegion(this, new TextureCoords(S0, T0, S1, T1) );
    }

    public void addReference() {
        refcount++;
    }

    public void removeReference() {
        refcount--;
        if (refcount<0)
            refcount = 0;
    }

    public boolean isReferenced() {
        return refcount>0;
    }

    public int getReferenceCount() {
        return refcount;
    }
		
	protected int mID;
	protected int width;
	protected int height;
	protected TextureCoords textCoords;
    protected int refcount = 0;
}

File: app/src/main/java/com/botijasoftware/utils/Texture.kt
package com.botijasoftware.utils

open class Texture {

    open var id: Int = 0
    var width: Int = 0
    var height: Int = 0
    var textureCoords: TextureCoords
        protected set
    var referenceCount = 0
        protected set

    constructor(id: Int, w: Int, h: Int) {
        this.id = id
        width = w
        height = h
        textureCoords = TextureCoords()
    }

    constructor(id: Int, w: Int, h: Int, tc: TextureCoords) {
        this.id = id
        width = w
        height = h
        textureCoords = tc
    }

    /*public Texture clone() {

        return new Texture(mID, width, height, textCoords.clone());

    }*/

    fun subTexture(S0: Float, T0: Float, S1: Float, T1: Float): TextureRegion {
        return TextureRegion(this, TextureCoords(S0, T0, S1, T1))
    }

    fun addReference() {
        referenceCount++
    }

    fun removeReference() {
        referenceCount--
        if (referenceCount < 0)
            referenceCount = 0
    }

    val isReferenced: Boolean
        get() = referenceCount > 0


}